### PR TITLE
[SPARK-15199] [SQL] Disallow Dropping Build-in Functions

### PR DESCRIPTION
**Pull request by gatorsmile**
#### What changes were proposed in this pull request?

As Hive and the major RDBMS behave, the built-in functions are not allowed to drop. In the current implementation, users can drop the built-in functions. However, after dropping the built-in functions, users are unable to add them back.
#### How was this patch tested?

Added a test case.

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @yhuai @andrewor14 @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala [hunk 1]
```diff
@@ -157,6 +157,9 @@ case class DropFunction(
         throw new AnalysisException(s"Specifying a database in DROP TEMPORARY FUNCTION " +
           s"is not allowed: '${databaseName.get}'")
       }
+      if (FunctionRegistry.builtin.functionExists(functionName)) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala]: what if users overwrite a built-in function and drop it?

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala]: I also think what you said is valid. If so, I think we also need to provide users a way to recover the built-in function, if they dropped it for any purpose. 

This PR is just to make our behavior consistent with Hive and the mainstream RDBMS. Normally, we do not allow users to drop the built-in functions. I am fine if we allow users to drop the built-in functions. 

Thanks!

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala]: Does hive allow users to overwrite built-in function?

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala]: Yeah, you can overwrite it by creating a temporary function, but you are unable to drop it. 

For example,
```
hive> drop function lower;
FAILED: SemanticException [Error 10301]: Cannot drop native function lower
hive> CREATE TEMPORARY FUNCTION lower AS 'org.apache.hadoop.hive.ql.udf.generic.GenericUDFUpper';
OK
Time taken: 0.005 seconds
hive> select lower('a');
OK
A
Time taken: 0.068 seconds, Fetched: 1 row(s)
hive> drop temporary function lower;
OK
Time taken: 0.01 seconds
hive> select lower('a');
OK
a
Time taken: 0.057 seconds, Fetched: 1 row(s)
hive> drop function lower;
FAILED: SemanticException [Error 10301]: Cannot drop native function lower
```

In this example, I overwrite the built-in function `lower` by creating the same name temporary function by the implementation of `upper`. Obviously, when I called `lower`, the actual logic is `upper`. That means, we can overwrite the built-in functions. After I dropping the temporary function, the built-in function becomes active again.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala]: Overwriting a built-in function is missing in the current implementation of the current `FunctionRegistry`. Should we support it in this release?

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala]: Thanks for the investigation! I guess we won't support this feature in 2.0, but need @yhuai to confirm.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala]: No problem. Just let me know if anything I can help. : )

**Comment by andrewor14** [comment]
LGTM merging into master 2.0. I confirmed with Yin offline.

**Pull request closed without merging**